providers/ms_entra_id.go

```diff
@@ -0,0 +1,166 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/util"
+	"github.com/spf13/cast"
+)
+
+// MicrosoftEntraIDProvider represents provider for Azure Entra Authentication V2 endpoint
+type MicrosoftEntraIDProvider struct {
+	*OIDCProvider
+	skipGraphGroups           bool
+	multiTenantAllowedTenants []string
+
+	microsoftGraphURL *url.URL
+	isMultiTenant     bool
+}
+
+const (
+	microsoftEntraIDProviderName = "Microsoft Entra ID"
+)
+
+var (
+	microsoftGraphURL = &url.URL{
+		Scheme: "https",
+		Host:   "graph.microsoft.com",
+		Path:   "/v1.0/me",
+	}
+)
+
+// NewAzureProvider initiates a new MicrosoftEntraIDProvider
+func NewMicrosoftEntraIDProvider(p *ProviderData, opts options.Provider) *MicrosoftEntraIDProvider {
+	p.setProviderDefaults(providerDefaults{
+		name: microsoftEntraIDProviderName,
+	})
+
+	return &MicrosoftEntraIDProvider{
+		OIDCProvider: NewOIDCProvider(p, opts.OIDCConfig),
+
+		skipGraphGroups:           opts.MicrosoftEntraIDConfig.DisableGroupsFromGraph,
+		multiTenantAllowedTenants: opts.MicrosoftEntraIDConfig.MultiTenantAllowedTenants,
+		isMultiTenant:             strings.Contains(opts.OIDCConfig.IssuerURL, "common"),
+		microsoftGraphURL:         microsoftGraphURL,
+	}
+}
+
+// EnrichSession checks for group overage and passes through to generic EnrichSession()
+func (p *MicrosoftEntraIDProvider) EnrichSession(ctx context.Context, session *sessions.SessionState) error {
+	err := p.OIDCProvider.EnrichSession(ctx, session)
+
+	if !p.skipGraphGroups {
+		logger.Printf("checking for overage of azure groups")
+		hasGroupOverage, _ := p.checkGroupOverage(session)
+
+		if hasGroupOverage {
+			logger.Printf("overage found, reading groups from graph")
+			if err = p.addGraphGroupsToSesion(ctx, session); err != nil {
+				return fmt.Errorf("unable to read groups from graph: %v", err)
+			}
+		}
+	}
+
+	return err
+}
+
+// ValidateSession checks for allowed tenants for multi-tenant apps and passes through to generic ValidateSession
+func (p *MicrosoftEntraIDProvider) ValidateSession(ctx context.Context, session *sessions.SessionState) bool {
+
+	if p.isMultiTenant {
+		issuer, exists, error := p.getIssuer(session)
+		if issuer == "" || !exists || error != nil {
+			return false
+		}
+
+		if p.multiTenantAllowedTenants != nil {
+			logger.Printf("List of allowed tenant is specified, we check if %s is allowed issuer", issuer)
+			tenantAllowed := p.checkIssuerMatchesTenantList(issuer, p.multiTenantAllowedTenants)
+			if !tenantAllowed {
+				return false
+			}
+		}
+	}
+
+	return p.OIDCProvider.ValidateSession(ctx, session)
+}
+
+// checkGroupOverage checks is claims present in ID Token indicate group overage
```

Comment:
```suggestion
// checkGroupOverage checks if the claims for the group membership are present in the ID Token
// https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages
```

Reply:
This is not exactly checking the group membership, but will fix.